homeassistant/components/somfy/__init__.py

```diff
@@ -0,0 +1,162 @@
+
+    hass.async_create_task(
+        hass.config_entries.flow.async_init(
+            DOMAIN,
+            context={'source': config_entries.SOURCE_IMPORT},
+        ))
+
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistantType, entry: ConfigEntry):
+    """Set up Somfy from a config entry."""
+    def token_saver(token):
+        _LOGGER.debug('Saving updated token')
+        entry.data[CONF_TOKEN] = token
+        hass.config_entries.async_update_entry(entry, data={**entry.data})
+
+    # Force token update.
+    from pymfy.api.somfy_api import SomfyApi
+    hass.data[DOMAIN][API] = SomfyApi(
+        entry.data['refresh_args']['client_id'],
+        entry.data['refresh_args']['client_secret'],
+        token=entry.data[CONF_TOKEN],
+        token_updater=token_saver
+    )
+
+    await update_all_devices(hass)
+
+    for component in SOMFY_COMPONENTS:
+        hass.async_create_task(
+            hass.config_entries.async_forward_entry_setup(
+                entry, component))
+
+    return True
+
+
+async def async_unload_entry(hass: HomeAssistantType, entry: ConfigEntry):
+    """Unload a config entry."""
+
+    if not hass.data[DOMAIN]:
+        hass.data.pop(DOMAIN)
+
+    for component in SOMFY_COMPONENTS:
+        await hass.config_entries.async_forward_entry_unload(
+            entry, component)
+
+    return True
+
+
+class SomfyEntity(Entity):
+    """Representation of a generic Somfy device."""
+
+    def __init__(self, device, api):
+        """Initialize the Somfy device."""
+        self.device = device
+        self.api = api
+
+    @property
+    def unique_id(self):
+        """Return the unique id base on the id returned by Somfy."""
+        return self.device.id
+
+    @property
+    def name(self):
+        """Return the name of the device."""
+        return self.device.name
+
+    @property
+    def device_info(self):
+        """Return device specific attributes.
+
+        Implemented by platform classes.
+        """
+        return {
+            'identifiers': {(DOMAIN, self.unique_id)},
+            'name': self.name,
+            'model': self.device.type,
+            'via_hub': (DOMAIN, self.device.site_id),
+        }
+
+    async def async_update(self):
+        """Update the device with the latest data."""
+        await update_all_devices(self.hass)
+        devices = self.hass.data[DOMAIN][DEVICES]
+        self.device = next((d for d in devices if d.id == self.device.id),
+                           self.device)
+
+    def has_capability(self, capability):
+        """Test if device has a capability."""
+        capabilities = self.device.capabilities
+        return bool([c for c in capabilities if c.name == capability])
+
+
+@Throttle(MIN_TIME_BETWEEN_UPDATES)
+async def update_all_devices(hass):
+    """Update all the devices."""
+    from requests import HTTPError
+    try:
+        data = hass.data[DOMAIN]
+        data[DEVICES] = data[API].get_devices()
```

Comment:
Is this async safe?

Reply:
`get_devices` does not perform any I/O action. It only call an URL: https://github.com/tetienne/somfy-open-api/blob/master/pymfy/api/somfy_api.py#L75

Reply:
A request to a url is I/O!

Reply:
Use hass.async_add_executor_job will be enough?

Reply:
Fixed.